src/precompiles/mod.rs

```diff
@@ -0,0 +1,73 @@
+use crate::{eravm_error::EraVmError, heaps::Heaps};
+use u256::U256;
+
+pub mod ecrecover;
+pub mod keccak256;
+pub mod secp256r1_verify;
+pub mod sha256;
+
+pub trait Precompile: std::fmt::Debug {
+    fn execute_precompile(&mut self, abi_key: U256, heaps: &mut Heaps) -> Result<(), EraVmError>;
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct MemoryQuery {
+    pub location: MemoryLocation,
+    pub value: U256,
+    pub rw_flag: bool,
+    pub value_is_pointer: bool,
+}
+
+impl MemoryQuery {
+    pub fn empty() -> Self {
+        Self {
+            location: MemoryLocation { page: 0, index: 0 },
+            value: U256::zero(),
+            rw_flag: false,
+            value_is_pointer: false,
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct MemoryLocation {
+    pub page: u32,
+    pub index: u32,
+}
```

Comment:
Can't we do this without reimplementing their way of accessing heap? I think the idea was to stop using this kind of access, and not just reimplementing it for our heap, if we have the page and offset we could just read from the heap with our current implementations.
But correct me if that is not the case